applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,888 @@
+            rElement.SetValue(WAKE_ELEMENTAL_DISTANCES, wake_elemental_distances);
+
+            // Save elemental distances in the nodes
+            for (unsigned int j = 0; j < wake_elemental_distances.size(); j++)
+            {
+                if (std::abs(wake_elemental_distances[j]) < mTolerance)
+                {
+                    if (wake_elemental_distances[j] < 0.0)
+                    {
+                        wake_elemental_distances[j] = - mTolerance;
+                    }
+                    else{
+                        wake_elemental_distances[j] = mTolerance;
+                    }
+                }
+                r_geometry[j].SetLock();
+                r_geometry[j].SetValue(WAKE_DISTANCE, wake_elemental_distances[j]);
+                r_geometry[j].UnSetLock();
+            }
+
+        }
+    });
+
+    // Add the trailing edge elements to the trailing_edge_sub_model_part
+    AddTrailingEdgeAndWakeElements(wake_elements_ordered_ids, trailing_edge_elements_ordered_ids);
+
+    KRATOS_INFO("MarkWakeElements") << "...Selecting wake elements finished..." << std::endl;
+    KRATOS_CATCH("");
+}
+
+// This function checks if the element is touching the trailing edge
+void Define3DWakeProcess::CheckIfTrailingEdgeElement(Element& rElement,
+                                                     Geometry<NodeType>& rGeometry,
+                                                     std::vector<std::size_t>& rTrailingEdgeElementsOrderedIds) const
+{
+    // Loop over element nodes
+    for (unsigned int i = 0; i < rGeometry.size(); i++)
+    {
+        // Elements touching the trailing edge are trailing edge elements
+        rGeometry[i].SetLock();
+        if (rGeometry[i].GetValue(TRAILING_EDGE))
+        {
+            rElement.SetValue(TRAILING_EDGE, true);
+            #pragma omp critical
+            {
+                rTrailingEdgeElementsOrderedIds.push_back(rElement.Id());
+            }
+        }
+        rGeometry[i].UnSetLock();
+    }
+}
+
+// This function adds the trailing edge elements in the
+// trailing_edge_sub_model_part
+void Define3DWakeProcess::AddTrailingEdgeAndWakeElements(std::vector<std::size_t>& rWakeElementsOrderedIds,
+                                                         std::vector<std::size_t>& rTrailingEdgeElementsOrderedIds) const
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+
+    std::sort(rWakeElementsOrderedIds.begin(),
+              rWakeElementsOrderedIds.end());
+    root_model_part.GetSubModelPart("wake_elements_model_part").AddElements(rWakeElementsOrderedIds);
+
+    std::sort(rTrailingEdgeElementsOrderedIds.begin(),
+              rTrailingEdgeElementsOrderedIds.end());
+    ModelPart& trailing_edge_sub_model_part =
+        root_model_part.GetSubModelPart("trailing_edge_elements_model_part");
+    trailing_edge_sub_model_part.AddElements(rTrailingEdgeElementsOrderedIds);
+
+    std::vector<std::size_t> trailing_edge_nodes_ordered_ids;
```

Comment:
maybe use reserve to improve speed (e.g. num_elements*3)

Reply:
Should the predicted reserved value be smaller or larger than the actual number of vector entries?

Reply:
depends:
if too small: reallocation (which is slow) will happen, but still this is usually done afaik
if too large: excess memory usage

Reply:
ok!